domain/network/state/types.go

```diff
@@ -0,0 +1,236 @@
+			}
+			snInfo.SetFan(underlay, overlay)
+		}
+
+		uniqueSpaces[space.UUID] = spInfo
+		uniqueSubnets[space.UUID][space.SubnetUUID] = snInfo
+
+		if _, ok := uniqueAZs[space.UUID]; !ok {
+			uniqueAZs[space.UUID] = make(map[string]map[string]string)
+		}
+		if _, ok := uniqueAZs[space.UUID][space.SubnetUUID]; !ok {
+			uniqueAZs[space.UUID][space.SubnetUUID] = make(map[string]string)
+		}
+		uniqueAZs[space.UUID][space.SubnetUUID][space.SubnetAZ] = space.SubnetAZ
+	}
+
+	// Iterate through every space and flatten its subnets.
+	for spaceUUID, space := range uniqueSpaces {
+		var subnets network.SubnetInfos
+		// Iterate through every subnet and flatten its availability zones.
+		for subnetUUID, subnet := range uniqueSubnets[spaceUUID] {
+			var availabilityZones []string
+			for _, availabilityZone := range uniqueAZs[spaceUUID][subnetUUID] {
+				availabilityZones = append(availabilityZones, availabilityZone)
+			}
+			subnet.AvailabilityZones = availabilityZones
+
+			subnets = append(subnets, subnet)
+		}
+		space.Subnets = subnets
+
+		res = append(res, space)
+	}
+
+	return res
+}
+
+// Subnet represents a single row from the database when
+// subnet is joined with subnet_type, subnet_association,
+// subject_subnet_type_uuid, subnet_association_type, provider_subnet,
+// provider_network, provider_network_subnet, availability_zone and
+// availability_zone_subnet.
+type Subnet struct {
+	// UUID is the subnet UUID.
+	UUID string `db:"uuid"`
+
+	// CIDR is the one of the subnet's cidr.
+	CIDR string `db:"cidr"`
+
+	// VLANTag is the subnet's vlan tag.
+	VLANTag int `db:"vlan_tag"`
+
+	// ProviderID is the subnet's provider id.
+	ProviderID string `db:"provider_id"`
+
+	// ProviderNetworkID is the subnet's provider network id.
+	ProviderNetworkID string `db:"provider_network_id"`
+
+	// ProviderSpaceUUID is the subnet's space uuid.
+	ProviderSpaceUUID sql.NullString `db:"provider_space_uuid"`
+
+	// SpaceUUID is the space uuid.
+	SpaceUUID sql.NullString `db:"space_uuid"`
+
+	// SpaceName is the name of the space the subnet is associated with.
+	SpaceName sql.NullString `db:"space_name"`
+
+	// AZ is the availability zones on the subnet.
+	AZ string `db:"az"`
+
+	// OverlayCIDR is the subnet's overlay cidr in a fan setup.
+	OverlayCIDR sql.NullString `db:"overlay_cidr"`
+
+	// UnderlayCIDR is the subnet's underlay cidr in a fan setup.
+	UnderlayCIDR sql.NullString `db:"underlay_cidr"`
+}
+
+// Alias type to a slice of Subnet('s).
+type Subnets []Subnet
+
+// ToSubnetInfos converts Subnets to a slice of network.SubnetInfo structs.
+// This method makes sure only unique AZs are mapped and flattens them into
+// each subnet.
+// No sorting is applied.
+func (sn Subnets) ToSubnetInfos() network.SubnetInfos {
+	var res network.SubnetInfos
+
+	// Prepare structs for unique subnets.
+	uniqueAZs := make(map[string]map[string]string)
+	uniqueSubnets := make(map[string]network.SubnetInfo)
+
+	for _, subnet := range sn {
+		sInfo := network.SubnetInfo{
```

Comment:
This section is largely duplicated in `ToSpaceInfos` and needs extraction.

Reply:
done in de1aeee9c3